std/bigint.d

```diff
@@ -1082,16 +1082,16 @@ Returns:
     A $(D string) that represents the $(D BigInt) as a decimal number.
 
 */
-string toDecimalString(const(BigInt) x)
+string toDecimalString(const(BigInt) x) pure
```

Comment:
This function should be `deprecated`. `myBigInt.to!string` is so much more idiomatic D.

Reply:
Problem is, that `BigInt.toString` uses the `delegate sink` method of `toString`, which makes it impossible to be `pure`.

I've been meaning to write up a bug report on how std.format should be reworked so that a possible `toString` overload of

```d
void toString(Writer)(auto ref Writer w) if (isOutputRange!(Writer, char))
```

is given precedence over every other `toString` overload.  The above signature allows all attributes to be inferred, along with fitting much better into the range paradigm.

Reply:
There is another way to do it. And that is, to pass the char buffer directly to toString, and let it work with that buffer. The only problem is, you need some way to communicate back to the caller if you don't have enough space.

I've thought of this since implementing iopipe, where I try to do everything directly in the buffer. The output range method works as well, but now the toString function is responsible for buffer management. For some things it works well, for others it may not.

The reason we opted (a long time ago) to use a char[] delegate instead of an output range is to allow virtual toString functions to work. Not sure it's as relevant today.

Reply:
There's even another: instead of returning a string, you return a range. If combined with a yielding Fiber, it's equally attractive syntaxwise and deals with the problem of passing the writer object around.

Reply:
> instead of returning a string, you return a range.

I think the toString version that receives an output range is more optimal, simply because then you can avoid dealing with buffering (if you don't need it), and the function could be implemented to take advantage of any optimizations based on the output range type. For example, if the output range given is an array, write directly to the array instead of using it as an output range.

Simple thought exercise, how would you return a toString range from an integer? Typically you construct the string from the right-most digit first! This does not work well in range form.

Reply:
Also, the main attraction of the output range overload (other than attributes) is that it allows the user to control the memory used.